api/coreservice.go

```diff
@@ -1381,26 +1382,70 @@ func (core *coreService) LogsInRange(filter *logfilter.LogFilter, start, end, pa
 		return nil, err
 	}
 
-	// TODO: improve using goroutine
 	if paginationSize == 0 {
 		paginationSize = 1000
 	}
 	if paginationSize > 5000 {
 		paginationSize = 5000
 	}
+
 	logs := []*iotextypes.Log{}
+	jobs := make(chan uint64, len(blockNumbers))
+	quit := make(chan error)
+	var wg sync.WaitGroup
+	logsMap := sync.Map{}
+	if len(blockNumbers) == 0 {
+		return logs, nil
+	}
 	for _, i := range blockNumbers {
-		logsInBlock, err := core.LogsInBlock(filter, i)
-		if err != nil {
+		jobs <- i
+	}
+	close(jobs)
+	for w := 1; w <= 5; w++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case err := <-quit:
```

Comment:
replace `quit` channel implementation with simple `errgroup.WithContext()`

Reply:
I can't make `errgroup` work functionally with `worker-pool`, needs some help, here's my current solution.
```go
logs := []*iotextypes.Log{}
jobs := make(chan uint64, len(blockNumbers))
logsMap := make(map[uint64][]*iotextypes.Log)
eg, ctx := errgroup.WithContext(context.Background())
if len(blockNumbers) == 0 {
    return logs, nil
}
for _, i := range blockNumbers {
    jobs <- i
}
close(jobs)
for w := 1; w <= 5; w++ {
    eg.Go(func() error {
        for {
		select {
		case <-ctx.Done():
			return nil
		default:
			job, ok := <-jobs
			if !ok {
				return errors.Errorf("invalid job number %d", job)
			}
			logsInBlock, err := core.logsInBlock(filter, job)
			if err != nil {
				return err
			}
			logsMap[job] = logsInBlock
		}
	}
    })
}
if err := eg.Wait(); err != nil {
    return nil, err
}

for _, i := range blockNumbers {
    logsInBlock := logsMap[i]
    for _, log := range logsInBlock {
	logs = append(logs, log)
	if len(logs) >= int(paginationSize) {
		return logs, nil
	}
    }
}
```

Reply:
And I haven't seen a working solution of `errgroup` + `worker-pool` by google search, is there any chance that these two are not suppose to use together?
The common solution is using `worker-pool` with `quit` channel, just like my previous commit.
Suggestion are welcome.

Reply:
see https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext

Reply:
Here is my code for reference: 
```
var (
		jobs      = make(chan jobDesc, len(blockNumbers))
		logsInBlk = make([][]*iotextypes.Log, len(blockNumbers))
		eg, ctx   = errgroup.WithContext(context.Background())
	)
	for i, v := range blockNumbers {
		jobs <- jobDesc{i, v}
	}
	close(jobs)
	for w := 1; w <= 10; w++ {
		eg.Go(func() error {
			for {
				select {
				case <-ctx.Done():
					return ctx.Err()
				default:
					job, ok := <-jobs
					if !ok {
						return nil
					}
					logsInBlock, err := core.logsInBlock(filter, job.blkNum)
					if err != nil {
						return err
					}
					logsInBlk[job.idx] = logsInBlock
				}
			}
		})
	}
	if err := eg.Wait(); err != nil {
		return nil, err
	}
```

Reply:
Thanks for the code and explanation.